File: sdk/go/receipt/rotation.go
package receipt

import (
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/pem"
	"fmt"
)

// algorithmEd25519 is the only signature algorithm the protocol supports
// (ADR-0001). Cross-algorithm rotation (e.g. Ed25519 → ML-DSA) is deferred to
// the algorithm-agility work and is rejected by verifiers until then.
const algorithmEd25519 = "ed25519"

// keyFingerprint returns the ADR-0015 fingerprint of a raw public key: the
// SHA-256 of the raw key bytes, rendered as sha256:<lowercase hex>. The raw
// bytes are the algorithm's canonical encoding (Ed25519: the 32-byte public
// key per RFC 8032 §5.1.5) — never an SPKI/PEM wrapper or a backend handle.
func keyFingerprint(raw []byte) string {
	sum := sha256.Sum256(raw)
	return "sha256:" + hex.EncodeToString(sum[:])
}

// decodeMultibaseEd25519Key decodes a multibase-"u" base64url string (the
// encoding ADR-0001 uses for proof.proofValue, applied here to raw public-key
// bytes) into a 32-byte Ed25519 public key.
func decodeMultibaseEd25519Key(s string) ([]byte, error) {
	if len(s) == 0 || s[0] != 'u' {
		return nil, fmt.Errorf("expected multibase %q prefix", "u")
	}
	raw, err := base64.RawURLEncoding.DecodeString(s[1:])
	if err != nil {
		return nil, fmt.Errorf("base64url decode: %w", err)
	}
	if len(raw) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("expected %d key bytes, got %d", ed25519.PublicKeySize, len(raw))
	}
	return raw, nil
}

// ed25519RawToPEM wraps a raw 32-byte Ed25519 public key in PEM-encoded SPKI,
// the form the signature verifier consumes.
func ed25519RawToPEM(raw []byte) (string, error) {
	der, err := x509.MarshalPKIXPublicKey(ed25519.PublicKey(raw))
	if err != nil {
		return "", fmt.Errorf("marshal SPKI public key: %w", err)
	}
	return string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der})), nil
}

// verifyRotationEvent validates the rotation-event fields of a key_rotated
// receipt against the outgoing (currently active) public key and returns the
// PEM-encoded incoming key that subsequent receipts must verify against.
//
// It implements the field-level checks of the ADR-0015 verifier traversal: the
// constant fields, the supported-algorithm guard, the old-key fingerprint
// consistency check against the outgoing key, and the new-key fingerprint check
// against the inline new_public_key. The rotation receipt's own signature is
// verified separately by the caller (it is signed with the outgoing key, so the
// chain verifier's normal per-receipt check covers it).
func verifyRotationEvent(activeKeyPEM string, kr *KeyRotation) (newKeyPEM string, err error) {
	if kr.EventType != "key_rotated" {
		return "", fmt.Errorf("event_type must be %q, got %q", "key_rotated", kr.EventType)
	}
	if kr.SignedWith != "old" {
		return "", fmt.Errorf("signed_with must be %q, got %q", "old", kr.SignedWith)
	}
	if kr.OldAlgorithm != algorithmEd25519 {
		return "", fmt.Errorf("unsupported old_algorithm %q: only %q is supported", kr.OldAlgorithm, algorithmEd25519)
	}
	if kr.NewAlgorithm != algorithmEd25519 {
		return "", fmt.Errorf("unsupported new_algorithm %q: only %q is supported", kr.NewAlgorithm, algorithmEd25519)
	}

	outRaw, err := parsePublicKey(activeKeyPEM)
	if err != nil {
		return "", fmt.Errorf("parse outgoing key: %w", err)
	}
	if got := keyFingerprint(outRaw); got != kr.OldKeyFingerprint {
		return "", fmt.Errorf("old_key_fingerprint mismatch: outgoing key is %s, field says %s", got, kr.OldKeyFingerprint)
	}

	newRaw, err := decodeMultibaseEd25519Key(kr.NewPublicKey)
	if err != nil {
		return "", fmt.Errorf("decode new_public_key: %w", err)
	}
	if got := keyFingerprint(newRaw); got != kr.NewKeyFingerprint {
		return "", fmt.Errorf("new_key_fingerprint mismatch: new_public_key hashes to %s, field says %s", got, kr.NewKeyFingerprint)
	}

	newKeyPEM, err = ed25519RawToPEM(newRaw)
	if err != nil {
		return "", fmt.Errorf("encode incoming key: %w", err)
	}
	return newKeyPEM, nil
}

File: sdk/ts/src/receipt/rotation.ts
import { createHash, createPublicKey } from "node:crypto";
import type { KeyRotation } from "./types.js";

/** The only signature algorithm the protocol supports (ADR-0001). Cross-algorithm
 *  rotation is deferred to the algorithm-agility work and rejected until then. */
const ALGORITHM_ED25519 = "ed25519";

/** Raw Ed25519 public keys are 32 bytes (RFC 8032 §5.1.5). */
const ED25519_PUBLIC_KEY_SIZE = 32;

/** ADR-0015 fingerprint of a raw public key: SHA-256 of the raw key bytes,
 *  rendered as sha256:<lowercase hex>. The bytes are the algorithm's canonical
 *  encoding — never an SPKI/PEM wrapper or a backend handle. */
export function keyFingerprint(raw: Buffer): string {
	return `sha256:${createHash("sha256").update(raw).digest("hex")}`;
}

/** Decode a multibase-"u" base64url string (the encoding ADR-0001 uses for
 *  proof.proofValue, applied here to raw public-key bytes) into a 32-byte
 *  Ed25519 public key. */
export function decodeMultibaseEd25519Key(s: string): Buffer {
	if (s.length === 0 || s[0] !== "u") {
		throw new Error('expected multibase "u" prefix');
	}
	if (!/^u[A-Za-z0-9_-]+$/.test(s)) {
		throw new Error("new_public_key is not multibase base64url");
	}
	const raw = Buffer.from(s.slice(1), "base64url");
	if (raw.length !== ED25519_PUBLIC_KEY_SIZE) {
		throw new Error(
			`expected ${ED25519_PUBLIC_KEY_SIZE} key bytes, got ${raw.length}`,
		);
	}
	return raw;
}

/** Wrap a raw 32-byte Ed25519 public key in PEM-encoded SPKI, the form the
 *  signature verifier consumes. */
export function ed25519RawToPem(raw: Buffer): string {
	const key = createPublicKey({
		key: { kty: "OKP", crv: "Ed25519", x: raw.toString("base64url") },
		format: "jwk",
	});
	return key.export({ type: "spki", format: "pem" }).toString();
}

/** Extract the raw 32-byte Ed25519 public key from a PEM-encoded SPKI key. */
export function pemToEd25519Raw(pem: string): Buffer {
	const jwk = createPublicKey(pem).export({ format: "jwk" });
	if (jwk.kty !== "OKP" || jwk.crv !== "Ed25519" || typeof jwk.x !== "string") {
		throw new Error("public key is not Ed25519");
	}
	return Buffer.from(jwk.x, "base64url");
}

export type RotationResult =
	| { ok: true; newKeyPem: string }
	| { ok: false; error: string };

/**
 * Validate the rotation-event fields of a key_rotated receipt against the
 * outgoing (currently active) public key and return the PEM-encoded incoming
 * key that subsequent receipts must verify against.
 *
 * Implements the field-level checks of the ADR-0015 verifier traversal: the
 * constant fields, the supported-algorithm guard, the old-key fingerprint
 * consistency check against the outgoing key, and the new-key fingerprint check
 * against the inline new_public_key. The rotation receipt's own signature is
 * verified separately by the caller (it is signed with the outgoing key).
 */
export function verifyRotationEvent(
	activeKeyPem: string,
	kr: KeyRotation,
): RotationResult {
	if (kr.event_type !== "key_rotated") {
		return {
			ok: false,
			error: `event_type must be "key_rotated", got "${kr.event_type}"`,
		};
	}
	if (kr.signed_with !== "old") {
		return {
			ok: false,
			error: `signed_with must be "old", got "${kr.signed_with}"`,
		};
	}
	if (kr.old_algorithm !== ALGORITHM_ED25519) {
		return {
			ok: false,
			error: `unsupported old_algorithm "${kr.old_algorithm}": only "${ALGORITHM_ED25519}" is supported`,
		};
	}
	if (kr.new_algorithm !== ALGORITHM_ED25519) {
		return {
			ok: false,
			error: `unsupported new_algorithm "${kr.new_algorithm}": only "${ALGORITHM_ED25519}" is supported`,
		};
	}

	let outRaw: Buffer;
	try {
		outRaw = pemToEd25519Raw(activeKeyPem);
	} catch (e) {
		return {
			ok: false,
			error: `parse outgoing key: ${e instanceof Error ? e.message : String(e)}`,
		};
	}
	const outFp = keyFingerprint(outRaw);
	if (outFp !== kr.old_key_fingerprint) {
		return {
			ok: false,
			error: `old_key_fingerprint mismatch: outgoing key is ${outFp}, field says ${kr.old_key_fingerprint}`,
		};
	}

	let newRaw: Buffer;
	try {
		newRaw = decodeMultibaseEd25519Key(kr.new_public_key);
	} catch (e) {
		return {
			ok: false,
			error: `decode new_public_key: ${e instanceof Error ? e.message : String(e)}`,
		};
	}
	const newFp = keyFingerprint(newRaw);
	if (newFp !== kr.new_key_fingerprint) {
		return {
			ok: false,
			error: `new_key_fingerprint mismatch: new_public_key hashes to ${newFp}, field says ${kr.new_key_fingerprint}`,
		};
	}

	try {
		return { ok: true, newKeyPem: ed25519RawToPem(newRaw) };
	} catch (e) {
		return {
			ok: false,
			error: `encode incoming key: ${e instanceof Error ? e.message : String(e)}`,
		};
	}
}
